### PR TITLE
Leaflet fixes

### PR DESCRIPTION
**Pull request by jasongrout**
This fixes some issues pointed out in #454. There still is some bug in the resizing I'm tracking down.

### CODE REVIEW

File: src/leafletwidget/widget.ts [hunk 1]
```diff
@@ -47,6 +51,25 @@ class MapWidget extends Widget {
...
+    // the size info. We can stop monkeypatching when we have a size hint change
+    // available from leaflet (cf.
+    // https://github.com/Leaflet/Leaflet/issues/4200#issuecomment-233616337 and
+    // https://github.com/jupyter/jupyterlab/pull/454#discussion_r71349224)
+    this._map.getSize = () => {
+      let map: any = this._map;
+      if (!map._size || map._sizeChanged) {
+        if (this._width < 0 || this._height < 0) {
+          return map.prototype.getSize.call(map);
+        } else {
```
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: no `else` when returning from `if` block

File: src/leafletwidget/widget.ts [hunk 2]
```diff
@@ -131,7 +145,9 @@ class MapWidget extends Widget {
     this.update();
   }
 
-  private _geojsonString = '';
+  private _width: number = -1;
```
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: `number` is the inferred type for `-1` value, so it's not strictly needed here.

Comment (jasongrout) [inline src/leafletwidget/widget.ts]: Thanks. I originally had null, but then didn't remove the type when I switched to -1 as the unknown option.

File: src/leafletwidget/widget.ts (hunk 1 above)
Comment (jasongrout) [inline src/leafletwidget/widget.ts]: as a style issue, to emphasize the return? I saw it as a style issue to emphasize the binary nature of the two branches. But I'm happy to defer - no strong opinions here.

Comment (jasongrout) [inline src/leafletwidget/widget.ts]: I suppose getting rid of the else means one level less of indentation, so +1.

Comment (sccolbert) [inline src/leafletwidget/widget.ts]: Yep. A readability issue. When you see `else` your mind is programmed to fall through the block, potentially missing the early exit condition.

Comment (sccolbert) [inline src/leafletwidget/widget.ts]: Which happened to me like 3 times reading this :)

**Comment by jasongrout** [comment]
@sccolbert - ready for review, as soon as the tests pass.

**Comment by jasongrout** [comment]
Ping @sccolbert for any other comments or merging...

File: src/leafletwidget/widget.ts [hunk 3]
```diff
@@ -80,47 +100,46 @@ class MapWidget extends Widget {
...
-    this._map.getSize = () => {
-      if (msg.width === -1 || msg.height === -1) {
-        return (this._map as any).prototype.getSize();
-      } else {
-        let size: any = new leaflet.Point(msg.width, msg.height);
-        (this._map as any)._size = size;
-        return size.clone();
-      }
-    };
+    this._width = msg.width;
```
Comment (blink1073) [inline src/leafletwidget/widget.ts]: How are you handling the case when `width` is `-1`?

Comment (sccolbert) [inline src/leafletwidget/widget.ts]: It's handled on line 63

File: src/leafletwidget/widget.ts [hunk 4]
```diff
@@ -47,9 +51,25 @@ class MapWidget extends Widget {
...
+        if (this._width < 0 || this._height < 0) {
+          return map.prototype.getSize.call(map);
+        }
+        map._size = new leaflet.Point(this._width, this._height);
+        map._sizeChanged = false;
+      }
+      return map._size.clone();
+    };
+
+    this.update();
```
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: Calling update from the constructor?

File: src/leafletwidget/widget.ts (hunk 3 above)
Comment (jasongrout) [inline src/leafletwidget/widget.ts]: Yep, I default back to the original getSize implementation, which does a bounding box read on the DOM node.

File: src/leafletwidget/widget.ts [hunk 5]
```diff
@@ -80,47 +100,46 @@ class MapWidget extends Widget {
...
     this._map.invalidateSize(true);
-    if (this._geojsonLayer) {
+    this._fitLayerBounds();
+  }
+
+  /**
+   * Make the map fit the geojson layer bounds only once when all info is available.
+   */
+  private _fitLayerBounds() {
+    if (this._fitBounds && this._geojsonLayer && this._width && this._height) {
```
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: this will be true if `_width` and `_height` are `-1`.

File: src/leafletwidget/widget.ts (hunk 4 above)
Comment (blink1073) [inline src/leafletwidget/widget.ts]: @jasongrout, we've typically called it in `onAfterAttach` if pertinent.

File: src/leafletwidget/widget.ts [hunk 6]
```diff
@@ -131,7 +150,10 @@ class MapWidget extends Widget {
     this.update();
   }
 
-  private _geojsonString = '';
+  private _fitBounds = true;
+  private _width = 0;
```
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: `-1` should be the initial size

File: src/leafletwidget/widget.ts (hunk 5 above)
Comment (jasongrout) [inline src/leafletwidget/widget.ts]: That's fine - then there is a DOM read. That handles the case when this is not inside a phosphor absolute-position layout.

File: src/leafletwidget/widget.ts (hunk 6 above)
Comment (jasongrout) [inline src/leafletwidget/widget.ts]: I want `_width` to initially be falsey so _fitLayerBounds only happens after a resize event (and I consider something that explicitly has width 0 as needing to still be fitted anyway...)

File: src/leafletwidget/widget.ts (hunk 4 above)
Comment (jasongrout) [inline src/leafletwidget/widget.ts]: Makes sense. Updating...

File: src/leafletwidget/widget.ts (hunk 5 above)
Comment (sccolbert) [inline src/leafletwidget/widget.ts]: Too much magic for my taste without explicit comments. But not my widget :)

Comment (jasongrout) [inline src/leafletwidget/widget.ts]: Good point, I'll put in a comment.

**Comment by sccolbert** [comment]
:shipit:

**Comment by jasongrout** [comment]
Too quick for me! Another PR coming up which mugglifies the code :).

**Pull request merged**